messaging/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -121,6 +121,10 @@ protected SimpleQueryBus(Builder builder) {
         this.spanFactory = builder.spanFactory;
     }
 
+    public <Q,R> boolean hasHandlersForMessage(QueryMessage<Q, R> query){
```

Comment:
I understand the introduction of this method, but I don't think that's a good idea, I'm afraid.

To know which handlers exist, I think it would be better if the `AxonServerQueryBus` would keep a list.
The reasoning behind this, is that the `AxonServerQueryBus` will subscribe all the handlers for the `localSegment` anyhow. Hence, all local handlers went through the `AxonServerQueryBus` to begin with.

By moving that logic to the `AxonServerQueryBus`, we maintain this feature in the `AxonServerQueryBus` as well, instead of spreading the concern.

Reply:
If I understood correctly, AxonServerQueryBus should keep its own list of local query handlers. This means both lists (the one inside of `AxonServerQueryBus `and `localSegment`) must be the same, otherwise delegating queryMessage from `AxonServerQueryBus `to `localSegemnt` can failed if a hander will be present on `AxonServerQueryBus` list but absent on `localSegment `list. Normally it shouldn't happend because the lists are created in `AxonServerQueryBus.subscirbe` method, but theoratically we could end up with two different list of local handers. right?

Reply:
That is indeed a correct assumption, @rsobies. Although, if the AF user would use Spring Boot or Axon Framework's configuration directly to subscribe query handlers, they'd always do so on the mean `QueryBus`, which is the `AxonServerQueryBus`.
The only scenario when this would go wrong is when a user deliberately retrieves the `localSegment` and invokes subscribe on it.

Although, indeed, theoretically an option, I took the stance with my suggestion that this chance is so slim, that it's worth to give it a try.

Regardless, good catch, @rsobies!

Reply:
>The only scenario when this would go wrong is when a user deliberately retrieves the localSegment and invokes subscribe on it.

I can see another scenario: user can invoke `cancel` from `Registration` which is return by `AxonServerQueryBus::subscribe` and the handler will be removed from `SimpleQueryBus` without notifying `AxonServerQueryBus`. In this case calling `AxonServerQueryBus::query` will throw NoHandlerForQueryException instead of sending the query to axon server.
I can't see any benefits for keeping additional list of registered local handlers in `AxonServerQueryBus`. Wouldn't be a better choice if we treat `localSegemnt` as single source of truth for query handlers?

Reply:
> I can see another scenario: user can invoke cancel from Registration which is return by AxonServerQueryBus::subscribe and the handler will be removed from SimpleQueryBus without notifying AxonServerQueryBus.

Ow, sharp yet again! But, this would mean the change should expand the `AxonServerQueryBus#subscribe` operation further, just so that the returned `Registration` also removes the subscribed handler from this additional list.

> Wouldn't be a better choice if we treat localSegemnt as single source of truth for query handlers?

My reasoning for not putting it on the `SimpleQueryBus`, is because I don't think the `public` method you added there as something Axon Framework should expose.

Reply:
>...without notifying AxonServerQueryBus.

but on other hand, it would be possible to put to `AxonServerRegistration` additional step for removing handler from the list, so both lists will be consistent, like this:

introducing new method in AxonServerQueryBus
private boolean unsubscribe(String queryName, Registration localSegmentRegistration)
{
        //todo: remove handler from axonServerQueryBus list
        return localSegmentRegistration.cancel();
 }

 
AxonServerQueryBus::subscribe {
....
return new AxonServerRegistration(() -> unsubscribe(queryName, localRegistration), serverRegistration::cancel);
}

Reply:
>My reasoning for not putting it on the SimpleQueryBus, is because I don't think the public method you added there as something Axon Framework should expose

ok, I see your point

Reply:
> I can't see any benefits for keeping additional list of registered local handlers in AxonServerQueryBus. Wouldn't be a better choice if we treat localSegemnt as single source of truth for query handlers?

Yes, it would be, if we can safely assume they all implementations expose this information. Unfortunately, we can't.

Since AxonServerQueryBus returns the Registration, we can also modify this Registration to also remove the information about the subscription from the "shortcut" list.